Introduced docs guide for emoji feature (#17713)
ckeditor/ckeditor5: packages/ckeditor5-emoji/docs/features/emoji.md

modified_at: 2025-01-03
---

{@snippet features/emoji-source}

# Emoji

The emoji feature lets you insert emojis to the document from the editor toolbar, or directly on the go while writing the content.

Since the number of emojis is very large, they have been conveniently divided into several categories for easier use, like "Smileys & Expressions" or "Objects". Besides, in order not to get lost, you may try to search for the emoji you are looking for using the search field.

## Demo

Use the emoji toolbar button {@icon @ckeditor/ckeditor5-emoji/theme/icons/emoji.svg Emoji} in the editor below to open a panel with a table of selectable emojis.

You may also trigger the emoji UI while writing the content when you type a pre-configured `:` marker followed by at least two letters of desired emoji name. Make sure to include at least one space before the marker, otherwise the emoji UI does not kick in. Then, you can either choose which emoji to insert from the suggestion list, or choose the last option &ndash; `Show all emoji...` &ndash; to open the full panel.

Additionally, there is also an option to set the skin tone of the emojis next to the search field.

> The selected emoji category and the tone skin are remembered by the feature, so the next time you invoke the panel you can quickly insert a similar emoji as before.

Reviewer: ```suggestion
The selected emoji category and the skin tone are remembered by the feature, so the next time you invoke the panel you can quickly insert an emoji of similar type.
```

{@snippet features/emoji}

<info-box info>
	This demo presents a limited set of features. Visit the {@link examples/builds/full-featured-editor feature-rich editor example} to see more in action.
</info-box>

> **Note:** Availability of emojis and their appearance depends on the operating system. If you want to standardize both the appearance and availability of emojis, please consider using an external font in your integration, such as [Noto Color Emoji](https://fonts.google.com/noto/specimen/Noto+Color+Emoji).

Reviewer: Maybe we should wrap this in `<info-box warning>`?

Author: Done in https://github.com/ckeditor/ckeditor5/pull/17713/commits/a420c9ca4b2e296d5156a9ee5c7b74731fac734a.

![obraz](https://github.com/user-attachments/assets/cffb92ff-6f4e-4052-a7d5-8afde0750910)

	The emoji feature requires the {@link features/mentions mentions feature} to be installed to work properly.
</info-box>

The {@link module:emoji/emoji~Emoji emoji plugin} consists of two sub-plugins:

* {@link module:emoji/emojimention~EmojiMention} &ndash; Adds support for triggering the emoji UI dropdown while writing the content when you type a pre-configured `:` marker followed by at least two letters of desired emoji name.
* {@link module:emoji/emojipicker~EmojiPicker} &ndash; Registers the UI button component which opens a panel with a table of selectable emojis.

> You can freely choose how to load the feature: either as a whole (as shown in the example above), or select one of the sub-plugins if you only need a specific way of inserting emojis into the document.

Reviewer: ```suggestion
You can freely choose how to load the feature: either as a whole (as shown in the example above), or select one of the sub-plugins if you need only one of the ways of inserting emojis into the document.
```

## Configuration

There is one configuration setting available for the emoji feature.

* `dropdownLimit` &ndash; the number of items to appear in the emoji dropdown.

```js
ClassicEditor
	.create( document.querySelector( '#editor' ), {
		// ... Other configuration options ...
		emoji: {
			dropdownLimit: 5
		}
	} )
	.then( /* ... */ )
	.catch( /* ... */ );
```

<info-box info>
	The emoji feature already has a pre-configured `:` marker that opens a panel with a table of selectable emojis. If you are using {@link features/mentions mentions} or {@link features/merge-fields merge fields} features, they can also show UI panels by pressing a pre-configured key and it may be in conflict with the emoji feature.

> 	To make all these features work fine together, make sure that {@link module:mention/mentionconfig~MentionFeed#marker mention's `marker`} and {@link module:merge-fields/mergefieldsconfig~MergeFieldsConfig#prefix merge field's `prefix`} configuration options are defined differently than `:`.

Reviewer: ```suggestion
	To prevent conflict between these features, make sure that {@link module:mention/mentionconfig~MentionFeed#marker mention's `marker`} and {@link module:merge-fields/mergefieldsconfig~MergeFieldsConfig#prefix merge field's `prefix`} configuration options are defined differently than `:`.
```

We need to make sure that we do not overlook this description in the future when we're adding other conflicting plugins. Also, do we have to also mention `Emoji` in feature guides for those plugins?

Author: IMHO I would keep it only here, because only emoji feature logs a warning about compatibility with other features.

</info-box>

## Related features

In addition to enabling emoji feature, you may want to check the following related features:

* {@link features/special-characters Special characters} &ndash; Allows inserting **mathematical operators**, **currency symbols**, **punctuation**, **graphic symbols** (such as arrows or bullets), or Unicode letters typically not accessible from the keyboard (such as **umlauts** or **other diacritics**).

> * {@link features/mentions Mentions} &ndash; Brings support for smart autocompletion.

Reviewer: I think that it does not make sense to put `Mention` here, as it is required by `Emoji` anyway.

Author: In this section we provide similar features and mention is very similar to emoji. Emoji underneath hack the mention plugin, but in my opinion it doesn't matter from the user's point of view. I would keep the link between mention and emoji features.

	The emoji feature requires the {@link features/mentions mentions feature} to be installed to work properly.
</info-box>

The {@link module:emoji/emoji~Emoji emoji plugin} consists of two sub-plugins:

* {@link module:emoji/emojimention~EmojiMention} &ndash; Adds support for triggering the emoji UI dropdown while writing the content when you type a pre-configured `:` marker followed by at least two letters of desired emoji name.
* {@link module:emoji/emojipicker~EmojiPicker} &ndash; Registers the UI button component which opens a panel with a table of selectable emojis.

You can freely choose how to load the feature: either as a whole (as shown in the example above), or select one of the sub-plugins if you need only one of the ways of inserting emojis into the document.

## Configuration

There is one configuration setting available for the emoji feature.

* `dropdownLimit` &ndash; the number of items to appear in the emoji dropdown.

```js
ClassicEditor
	.create( document.querySelector( '#editor' ), {
		// ... Other configuration options ...
		emoji: {
			dropdownLimit: 5
		}
	} )
	.then( /* ... */ )
	.catch( /* ... */ );
```

<info-box info>

> 	The emoji feature already has a pre-configured `:` marker that opens a panel with a table of selectable emojis. If you are using {@link features/mentions mentions} or {@link features/merge-fields merge fields} features, they can also show UI panels by pressing a pre-configured key and it may be in conflict with the emoji feature.

Reviewer 2: ```suggestion
	The emoji feature uses the `:` marker that opens a panel with a table of selectable emojis. If you are using {@link features/mentions mentions} or {@link features/merge-fields merge fields} features, they can also show UI panels by pressing a pre-configured key, and it may conflict with the emoji feature. In such a case, the {@link module:emoji/emojimention~EmojiMention} plugin will not integrate the autocompletion mechanism.
```

I would add a note of what happens when a conflict occurs.

> 	To prevent conflict between these features, make sure that {@link module:mention/mentionconfig~MentionFeed#marker mention's `marker`} and {@link module:merge-fields/mergefieldsconfig~MergeFieldsConfig#prefix merge field's `prefix`} configuration options are defined differently than `:`.

Reviewer 2: ```suggestion
	To prevent conflicts, make sure that {@link module:mention/mentionconfig~MentionFeed#marker mention's `marker`} and {@link module:merge-fields/mergefieldsconfig~MergeFieldsConfig#prefix merge field's `prefix`} configuration options are defined differently than `:`.
```

Could be simplified, a bit.
